src/dbus/dbus_thread_object.cpp

```diff
@@ -0,0 +1,558 @@
+#endif
+        threadHelper->Attach(name, panid, extPanId, masterKey, pskc, channelMask, [aRequest](otError aError) mutable {
+            aRequest.ReplyOtbrResult(static_cast<otbrError>(aError));
+        });
+    }
+}
+
+void DBusThreadObject::FactoryResetHandler(DBusRequest &aRequest)
+{
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+    mNcp->Reset();
+
+    {
+        auto threadHelper = mNcp->GetThreadHelper();
+        threadHelper->AddDeviceRoleHandler(std::bind(&DBusThreadObject::DeviceRoleHandler, this, _1));
+        SignalPropertyChanged(OTBR_DBUS_PREFIX, OTBR_DBUS_DEVICE_ROLE_PROPERTY,
+                              GetDeviceRoleName(OT_DEVICE_ROLE_DISABLED));
+    }
+}
+
+void DBusThreadObject::ResetHandler(DBusRequest &aRequest)
+{
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    otInstanceReset(threadHelper->GetInstance());
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+}
+
+void DBusThreadObject::JoinerStartHandler(DBusRequest &aRequest)
+{
+    auto        threadHelper = mNcp->GetThreadHelper();
+    std::string pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData;
+    auto        args = std::tie(pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+        threadHelper->JoinerStart(pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData,
+                                  [aRequest](otError aError) mutable { aRequest.ReplyOtbrResult(OTBR_ERROR_NONE); });
+    }
+}
+
+void DBusThreadObject::JoinerStopHandler(DBusRequest &aRequest)
+{
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    otJoinerStop(threadHelper->GetInstance());
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+}
+
+void DBusThreadObject::OpenUnsecurePortHandler(DBusRequest &aRequest)
+{
+    auto     threadHelper = mNcp->GetThreadHelper();
+    uint16_t port;
+    uint32_t timeout;
+    auto     args = std::tie(port, timeout);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+        aRequest.ReplyOtbrResult(static_cast<otbrError>(threadHelper->OpenUnsecurePort(port, timeout)));
+    }
+}
+
+void DBusThreadObject::AddOnMeshPrefixHandler(DBusRequest &aRequest)
+{
+    auto             threadHelper = mNcp->GetThreadHelper();
+    otbrOnMeshPrefix onMeshPrefix;
+    auto             args = std::tie(onMeshPrefix);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+        otBorderRouterConfig cfg;
+
+        // size is guaranteed by parsing
+        memcpy(&cfg.mPrefix.mPrefix, &onMeshPrefix.mPrefix.mPrefix, sizeof(cfg.mPrefix.mPrefix));
```

Comment:
`onMeshPrefix.mPrefix.mPrefix` is a `std::vector`.

We should not use `memcpy` if possible.

Reply:
cannot find the definition of `otBorderRouterConfig`. Can we do assignment here?

Reply:
the `mLength` field of `otbrIp6Prefix` should be eliminated.

Reply:
The mLength field doesn't mean the byte length of a prefix. The definition is the same as `otIp6Prefix`. The mLength for `ff02::2` will be 2.

Reply:
@gjc13 we are using `memcpy`  on `onMeshPrefix.mPrefix.mPrefix`, which is not POD.